rcl/src/rcl/node.c

```diff
@@ -542,6 +564,173 @@ rcl_get_disable_loaned_message(bool * disable_loaned_message)
+      &response)))
+  {
+    RCUTILS_LOG_ERROR_NAMED(
+      ROS_PACKAGE_NAME,
+      "Failed to initialize service request / response.");
+    goto cleanup;
+  }
+
+  if (RCL_RET_OK != rcl_take_request(
+      &node->impl->get_type_description_service,
+      &request_header, &request))
+  {
+    RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME, "Failed to take request");
+    goto cleanup;
+  }
+
+  rosidl_type_hash_t type_hash;
+  if (RCUTILS_RET_OK !=
+    rosidl_parse_type_hash_string(request.type_hash.data, &type_hash))
+  {
+    RCUTILS_LOG_ERROR_NAMED(
+      ROS_PACKAGE_NAME, "Failed to parse type hash '%s'",
+      request.type_hash.data);
+  }
+
+  rcl_ret_t ret =
+    rcl_node_type_cache_get_type_info(node, &type_hash, &type_info);
+
+  if (RCUTILS_RET_OK == ret) {
+    response.successful = type_description_interfaces__msg__TypeDescription__copy(
+      type_info.type_description, &response.type_description);
+
+    if (request.include_type_sources) {
+      response.successful &= type_description_interfaces__msg__TypeSource__Sequence__copy(
+        type_info.type_sources, &response.type_sources);
+    }
+
+    if (!response.successful) {
+      rosidl_runtime_c__String__assign(
+        &response.failure_reason,
+        "Failed to populate response");
+    }
+  } else {
+    response.successful = false;
+    RCUTILS_LOG_INFO_NAMED(
+      ROS_PACKAGE_NAME,
+      "Type '%s' not found in type cache",
+      request.type_hash.data);
+    rosidl_runtime_c__String__assign(
+      &response.failure_reason,
+      "Type not currently used by this node");
+  }
+
+  if (RCL_RET_OK != rcl_send_response(
+      &node->impl->get_type_description_service,
+      &request_header, &response))
+  {
+    RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME, "Failed to send response");
+  }
+
+cleanup:
+  type_description_interfaces__srv__GetTypeDescription_Request__fini(&request);
+  type_description_interfaces__srv__GetTypeDescription_Response__fini(
+    &response);
+}
+
+rcl_ret_t rcl_node_type_description_service_init(rcl_node_t * node)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(node, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(node->impl, RCL_RET_NODE_INVALID);
+
+  if (rcl_service_is_valid(&node->impl->get_type_description_service)) {
+    return RCL_RET_ALREADY_INIT;
+  }
+  rcl_reset_error();  // Reset the error message set by rcl_service_is_valid()
+
+  char * serviceName = NULL;
+  const rosidl_service_type_support_t * type_support =
+    ROSIDL_GET_SRV_TYPE_SUPPORT(
+    type_description_interfaces, srv,
+    GetTypeDescription);
+  rcl_service_options_t service_ops = rcl_service_get_default_options();
+  rcl_allocator_t allocator = node->context->impl->allocator;
+
+  // Construct service name
+  rcl_ret_t ret = rcl_node_resolve_name(
+    node, "~/get_type_description",
+    allocator, true, true, &serviceName);
+  if (RCL_RET_OK != ret) {
+    RCL_SET_ERROR_MSG(
+      "Failed to construct ~/get_type_description service name");
+    goto cleanup;
```

Comment:
I think we can just return here, as if this failed we haven't allocated any memory up to this point.

Reply:
```suggestion
    return ret;
```